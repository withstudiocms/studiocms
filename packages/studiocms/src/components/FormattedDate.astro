---
import config from "studiocms:config";

interface Props {
	date: Date;
}

const datetime = Astro.props.date.toISOString();
const formattedDate = Astro.props.date.toLocaleDateString(
	config.locale.dateLocale,
	config.locale.dateTimeFormat,
);
---

<time {datetime}>{formattedDate}</time>
