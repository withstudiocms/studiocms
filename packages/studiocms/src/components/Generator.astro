---
---
<meta name="generator" content={Astro.generator} />
<meta name="generator" content={Astro.locals.SCMSGenerator} />
<meta name="generator" content={Astro.locals.SCMSUiGenerator} />