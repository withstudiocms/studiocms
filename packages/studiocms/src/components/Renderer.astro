---
import contentRenderer from 'studiocms:renderer/current';

interface Props {
	content: string;
}

const { content } = Astro.props;

// @ts-ignore
const renderedContent = await contentRenderer(content, $$result);
---
<Fragment set:html={renderedContent}></Fragment>