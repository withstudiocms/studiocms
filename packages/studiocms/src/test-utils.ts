import type { AstroIntegration, AstroIntegrationLogger } from 'astro';
import type { StudioCMSPlugin } from './plugins.js';
import type {
	SCMSAuthServiceFnOpts,
	SCMSDashboardFnOpts,
	SCMSFrontendFnOpts,
	SCMSImageServiceFnOpts,
	SCMSRenderingFnOpts,
	SCMSSiteMapFnOpts,
} from './schemas/index.js';

type HookRun<T> = { hasHook: boolean; hookResults: T };
export interface PluginHookResults {
	astroConfig: HookRun<{ integrations: AstroIntegration[] }>;
	studiocmsConfig: HookRun<{
		authService: Partial<SCMSAuthServiceFnOpts>;
		dashboard: Partial<SCMSDashboardFnOpts>;
		frontend: Partial<SCMSFrontendFnOpts>;
		imageService: Partial<SCMSImageServiceFnOpts>;
		rendering: Partial<SCMSRenderingFnOpts>;
		sitemap: Partial<SCMSSiteMapFnOpts>;
	}>;
}
/**
 * Utility class for testing StudioCMS plugins by simulating hook execution and collecting results.
 *
 * The `StudioCMSPluginTester` provides methods to:
 * - Create mock loggers for use in plugin hooks.
 * - Execute the `studiocms:astro:config` and `studiocms:config:setup` hooks, if present, on a given plugin.
 * - Collect and return integration and configuration data set by these hooks.
 * - Retrieve basic plugin metadata.
 *
 * @remarks
 * This class is intended for use in test environments to facilitate inspection and validation of plugin behavior.
 *
 * @example
 * ```typescript
 * const tester = new StudioCMSPluginTester(myPlugin);
 * const info = tester.getPluginInfo();
 * const hookResults = await tester.getHookResults();
 * ```
 */
export class StudioCMSPluginTester {
	private plugin: StudioCMSPlugin;
	private readonly injectedLogger?: AstroIntegrationLogger;

	constructor(plugin: StudioCMSPlugin, logger?: AstroIntegrationLogger) {
		this.plugin = plugin;
		this.injectedLogger = logger;
	}

	/**
	 * Creates a mock logger object for testing purposes.
	 *
	 * The returned logger implements the `AstroIntegrationLogger` interface and provides
	 * stubbed logging methods (`info`, `warn`, `error`, `debug`) that delegate to the corresponding
	 * `console` methods. It also includes a `fork` method, which returns a new mock logger
	 * instance with an attached label for testing logger forking behavior.
	 *
	 * @returns {AstroIntegrationLogger} A mock logger instance suitable for use in tests.
	 */
	private createMockLogger(): AstroIntegrationLogger {
		if (this.injectedLogger) return this.injectedLogger;
		// biome-ignore lint/suspicious/noExplicitAny: this is fine
		const logger: Record<string, any> = {
			info: console.log,
			warn: console.warn,
			error: console.error,
			debug: console.debug,
			fork: (label: string) => {
				// Each fork returns a new mock logger with the label attached for testing
				const forked = this.createMockLogger();
				forked.label = label;
				return forked;
			},
		};
		return logger as unknown as AstroIntegrationLogger;
	}

	/**
	 * Executes the 'studiocms:astro:config' hook if it exists on the plugin,
	 * providing a mock logger and a method to collect Astro integrations.
	 *
	 * @returns A promise that resolves to an object containing the collected integrations.
	 *
	 * @remarks
	 * This method checks if the plugin defines a 'studiocms:astro:config' hook as a function.
	 * If so, it invokes the hook with a mock logger and an `addIntegrations` callback,
	 * which accumulates integrations into an array. The collected integrations are then
	 * returned in an object.
	 */
	private async runAstroConfigHook() {
		const integrations: AstroIntegration[] = [];
		if (
			this.plugin.hooks['studiocms:astro:config'] &&
			typeof this.plugin.hooks['studiocms:astro:config'] === 'function'
		) {
			await this.plugin.hooks['studiocms:astro:config']({
				logger: this.createMockLogger(),
				addIntegrations: (newIntegrations) => {
					const toAdd = Array.isArray(newIntegrations) ? newIntegrations : [newIntegrations];
					integrations.push(...toAdd);
				},
			});
		}
		return { integrations };
	}

	/**
	 * Executes the `studiocms:config:setup` hook on the current plugin, if available,
	 * and collects configuration options set by the hook into partial option objects.
	 *
	 * This method mocks the hook's context by providing setter functions for various
	 * StudioCMS configuration aspects, such as authentication, dashboard, frontend,
	 * image service, rendering, and sitemap. Each setter captures the values provided
	 * by the hook and stores them in corresponding partial option objects.
	 *
	 * @returns A promise that resolves to an object containing the collected configuration
	 *          options for authentication service, dashboard, frontend, image service,
	 *          rendering, and sitemap.
	 *
	 * @remarks
	 * This utility is primarily intended for testing purposes, allowing inspection of
	 * configuration values set by a plugin's `studiocms:config:setup` hook.
	 */
	private async runStudioCMSConfigHook() {
		const authService: Partial<SCMSAuthServiceFnOpts> = {};
		const dashboard: Partial<SCMSDashboardFnOpts> = {};
		const frontend: Partial<SCMSFrontendFnOpts> = {};
		const imageService: Partial<SCMSImageServiceFnOpts> = {};
		const rendering: Partial<SCMSRenderingFnOpts> = {};
		const sitemap: Partial<SCMSSiteMapFnOpts> = {};

		if (
			this.plugin.hooks['studiocms:config:setup'] &&
			typeof this.plugin.hooks['studiocms:config:setup'] === 'function'
		) {
			await this.plugin.hooks['studiocms:config:setup']({
				logger: this.createMockLogger(),
				setAuthService: ({ oAuthProvider }) => {
					if (oAuthProvider) {
						authService.oAuthProvider = oAuthProvider;
					}
				},
				setDashboard: ({ dashboardGridItems, dashboardPages }) => {
					if (dashboardGridItems) {
						dashboard.dashboardGridItems = dashboardGridItems;
					}
					if (dashboardPages) {
						dashboard.dashboardPages = dashboardPages;
					}
				},
				setFrontend: ({ frontendNavigationLinks }) => {
					if (frontendNavigationLinks) {
						frontend.frontendNavigationLinks = frontendNavigationLinks;
					}
				},
				setImageService: ({ imageService: imgService }) => {
					if (imgService) {
						imageService.imageService = imgService;
					}
				},
				setRendering: ({ pageTypes }) => {
					if (pageTypes) {
						rendering.pageTypes = pageTypes;
					}
				},
				setSitemap: ({ sitemaps, triggerSitemap }) => {
					if (sitemaps) {
						sitemap.sitemaps = sitemaps;
					}
					sitemap.triggerSitemap = triggerSitemap;
				},
			});
		}

		return {
			authService,
			dashboard,
			frontend,
			imageService,
			rendering,
			sitemap,
		};
	}

	/**
	 * Retrieves information about the current plugin.
	 *
	 * @returns An object containing the plugin's identifier, name, minimum required StudioCMS version, and dependencies.
	 */
	public getPluginInfo() {
		return {
			identifier: this.plugin.identifier,
			name: this.plugin.name,
			studiocmsMinimumVersion: this.plugin.studiocmsMinimumVersion,
			requires: this.plugin.requires,
		};
	}

	/**
	 * Asynchronously retrieves the results of configured plugin hooks.
	 *
	 * @returns An object containing the presence and results of the 'studiocms:astro:config' and 'studiocms:config:setup' hooks.
	 * - `astroConfig`: Indicates if the 'studiocms:astro:config' hook exists and provides its execution results.
	 * - `studiocmsConfig`: Indicates if the 'studiocms:config:setup' hook exists and provides its execution results.
	 */
	public async getHookResults(): Promise<PluginHookResults> {
		return {
			astroConfig: {
				hasHook: !!this.plugin.hooks['studiocms:astro:config'],
				hookResults: await this.runAstroConfigHook(),
			},
			studiocmsConfig: {
				hasHook: !!this.plugin.hooks['studiocms:config:setup'],
				hookResults: await this.runStudioCMSConfigHook(),
			},
		};
	}
}
