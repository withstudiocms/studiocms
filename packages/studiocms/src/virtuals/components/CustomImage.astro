---
/// <reference types="astro/client" />
import { Image } from 'astro:assets';
import config from 'studiocms:config';
import * as imageServices from 'studiocms:plugins/imageService';
import { imageServiceKeys } from 'studiocms:plugins/imageService';
import type { StudioCMSImageService } from '../../schemas/index.js';

const {
	features: { preferredImageService },
} = config;

interface Props {
	src: string;
	alt: string;
	width: number;
	height: number;
}

const { src, ...props } = Astro.props;

// Get the current configured image service
async function getImageSrc() {
	const keys = imageServiceKeys;

	if (keys.length === 0 || !preferredImageService) {
		return src;
	}

	/* v8 ignore start */
	const preferredService = keys.find(({ identifier }) => {
		return identifier === preferredImageService;
	});

	if (!preferredService) {
		return src;
	}

	// @ts-expect-error - This is a dynamic type... from a virtual module
	const serviceFn = imageServices[preferredService.safe] as unknown as StudioCMSImageService;

	if (typeof serviceFn !== 'function') {
		console.warn(`Image service "${preferredImageService}" is not properly configured`);
		return src;
	}

	return await serviceFn(src, props);
}
/* v8 ignore stop */

// check if the image source is an external URL
function inferSizeNeeded(url: string) {
	if (url.startsWith('https://') || url.startsWith('http://')) {
		return true;
	}
	/* v8 ignore start */
	return false;
}
/* v8 ignore stop */

// Get the current configured image service source URL
const currentSrcURL = await getImageSrc();

// Set the image properties
const imageProps = {
	src: currentSrcURL,
	inferSize: inferSizeNeeded(currentSrcURL),
	...props,
};
---

<Image loading={"lazy"} {...imageProps} />
