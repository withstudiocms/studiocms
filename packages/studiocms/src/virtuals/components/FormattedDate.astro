---
import config from 'studiocms:config';

const {
	locale: { dateLocale: locales, dateTimeFormat: options },
} = config;

interface Props {
	date: Date;
	__test_mode?: boolean;
}

const { date, __test_mode = false } = Astro.props;

const datetime = date.toISOString();
const formattedDate = __test_mode ? date.toUTCString() : date.toLocaleDateString(locales, options);
---

<time {datetime}>{formattedDate}</time>
