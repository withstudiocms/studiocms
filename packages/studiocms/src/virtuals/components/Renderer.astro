---
import { renderAugments } from 'studiocms:plugins/augments';
import { pluginRenderers } from 'studiocms:plugins/renderers';
import * as renderers from 'virtual:studiocms/plugins/renderers';
import type { SSRResult } from 'astro';
import type { PluginPageTypeRendererProps, PluginRenderer } from '../../types.js';
import renderFn from './renderFn.js';

interface Props extends PluginPageTypeRendererProps {}

// Get props
const { data } = Astro.props as Props;

// If no data, return nothing
if (!data) return;

// Find the renderer for the given page type
const renderer = pluginRenderers.find((r) => r.pageType === data.package);

/* v8 ignore start */
if (!renderer) {
	console.warn(
		`Renderer for page type "${data.package}" not found. Ensure the plugin is correctly registered and the renderer exists.`
	);
	// If no renderer found, return nothing
	return;
}
/* v8 ignore stop */

// Get the renderer module
const RendererModule: PluginRenderer = renderers[renderer.safePageType as keyof typeof renderers];

/* v8 ignore start */
// Check if RendererModule has a name property to verify it's a valid interface
if (!RendererModule.name) {
	console.error(`Renderer module for page type "${data.package}" not found in renderers.`);
	// If no renderer module found, return nothing
	return;
}
/* v8 ignore stop */

// Filter augments to only include those specified in the page data
const augments = renderAugments.filter((augment) =>
	(data.augments as string[])?.includes(augment.id)
);

// Render the content using the render function
// If no content found, return an error message
// Pass in the preRenderer and sanitizeOpts if they exist on the RendererModule
const renderedContent = await renderFn({
	content: data.defaultContent?.content || '<h1>Error: No content found</h1>',
	augments,
	renderOpts: {
		// @ts-expect-error - $$result is a global variable injected by Astro during compilation and not a editor-known variable
		result: $$result as SSRResult,
		sanitizeOpts: RendererModule.sanitizeOpts,
		preRenderer: RendererModule.renderer,
	},
});
---

<Fragment set:html={renderedContent} />
