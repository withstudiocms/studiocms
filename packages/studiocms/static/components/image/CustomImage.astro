---
/// <reference types="astro/client" />
import { Image } from 'astro:assets';
import { getSecret } from 'astro:env/server';
import Config from 'studiocms:config';
import { Cloudinary } from '@cloudinary/url-gen';
import { fill } from '@cloudinary/url-gen/actions/resize';

const CMS_CLOUDINARY_CLOUDNAME = getSecret('CMS_CLOUDINARY_CLOUDNAME');

interface Props {
	src: string;
	alt: string;
	width: number;
	height: number;
}

/**
 * Cloudinary Plugin
 *
 * This plugin is used to generate Cloudinary URLs for images using `@cloudinary/url-gen` for the StudioCMS `CustomImage` component.
 *
 * @param src the image name or URL
 * @param props the image props
 * @returns the Cloudinary image URL for the given source and props
 */
export function cloudinaryPlugin(src: string, props: Omit<Props, 'src'>) {
	// Cloudinary Image Service (JavaScript SDK) - https://cloudinary.com/documentation/javascript_integration#landingpage

	// Initialize Cloudinary
	const cld = new Cloudinary({
		cloud: {
			cloudName: CMS_CLOUDINARY_CLOUDNAME || '',
		},
	});

	// Configure the image
	const cldSrc = cld
		.image(src)
		.format('auto')
		.quality('auto')
		.resize(fill('auto').width(props.width).height(props.height));

	// Set the delivery type to 'fetch' if the image source is an external URL
	if (src.startsWith('https://') || src.startsWith('http://')) {
		cldSrc.setDeliveryType('fetch');
	}

	// Return the Cloudinary image URL
	return cldSrc.toURL();
}

const { src, ...props } = Astro.props;

// Get the current configured image service
const {
	imageService: { cdnPlugin },
} = Config;

// Get the current configured image service
async function getImageSrc() {
	switch (cdnPlugin) {
		case 'cloudinary-js':
			// Return the Cloudinary image source
			return cloudinaryPlugin(src, props);
		default:
			// Return the original image source if no External image service plugin is configured
			return src;
	}
}

// check if the image source is an external URL
function inferSizeNeeded(url: string) {
	if (url.startsWith('https://') || url.startsWith('http://')) {
		return true;
	}
	return false;
}

// Get the current configured image service source URL
const currentSrcURL = await getImageSrc();

// Set the image properties
const imageProps = {
	src: currentSrcURL,
	inferSize: inferSizeNeeded(currentSrcURL),
	...props,
};
---
<Image loading={"lazy"} {...imageProps}/>