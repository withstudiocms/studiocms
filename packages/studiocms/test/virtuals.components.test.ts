/// <reference types="astro/client" />
import { describe, expect } from 'vitest';
import CustomImage from '../src/virtuals/components/CustomImage.astro';
import FormattedDate from '../src/virtuals/components/FormattedDate.astro';
import Generator from '../src/virtuals/components/Generator.astro';
import Renderer from '../src/virtuals/components/Renderer.astro';
import { test } from './fixtures/AstroContainer';
import { MockAstroLocals, makeRendererProps } from './test-utils';

describe('Virtual Components Container tests', () => {
	describe('CustomImage component', () => {
		test('render component', async ({ renderComponent }) => {
			const result = await renderComponent(CustomImage, 'CustomImage', {
				props: { src: 'https://seccdn.libravatar.org/static/img/mm/80.png', alt: 'Test Image' },
			});
			expect(result).toMatchSnapshot();
		});
	});

	describe('FormattedDate component', () => {
		test('render component', async ({ renderComponent }) => {
			const result = await renderComponent(FormattedDate, 'FormattedDate', {
				props: { date: new Date('2020-01-01'), __test_mode: true },
			});
			expect(result).toMatchSnapshot();
		});
	});

	describe('Generator component', () => {
		test('render component', async ({ renderComponent }) => {
			const result = await renderComponent(Generator, 'Generator', {
				props: { locals: MockAstroLocals() },
			});
			expect(result).toMatchSnapshot();
		});
	});

	describe('Renderer component', () => {
		test('render component with content', async ({ renderComponent }) => {
			const result = await renderComponent(Renderer, 'Renderer', {
				props: makeRendererProps(
					'<div>This is a test content from the defaultContent field.</div>'
				),
			});
			expect(result).toMatchSnapshot();
		});

		test('render component with no content', async ({ renderComponent }) => {
			const result = await renderComponent(Renderer, 'Renderer', {
				props: makeRendererProps(null),
			});
			expect(result).toMatchSnapshot();
		});
	});
});
