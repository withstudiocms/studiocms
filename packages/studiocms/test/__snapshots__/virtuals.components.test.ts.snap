// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Virtual Components Container tests > CustomImage component > render component 1`] = `"<img src="https://seccdn.libravatar.org/static/img/mm/80.png" data-image-component="true" loading="lazy" alt="Test Image" decoding="async" fetchpriority="auto" width="80" height="80">"`;

exports[`Virtual Components Container tests > FormattedDate component > render component 1`] = `"<time datetime="2020-01-01T00:00:00.000Z">Wed, 01 Jan 2020 00:00:00 GMT</time>"`;

exports[`Virtual Components Container tests > Generator component > render component 1`] = `"<meta name="generator" content="Astro v0.0.0-test"><meta name="generator" content="StudioCMS v0.0.0-test"><meta name="generator" content="StudioCMS UI v0.0.0-test">"`;

exports[`Virtual Components Container tests > Renderer component > render component with content 1`] = `"<div>This is a test content from the defaultContent field.</div>"`;

exports[`Virtual Components Container tests > Renderer component > render component with no content 1`] = `"<h1>Error: No content found</h1>"`;

exports[`Virtual Components Container tests > i18n Language Selector component > render component 1`] = `"<div class="sui-dropdown-container" data-align="start" id="language-selector-container" data-trigger="left"> <div class="sui-dropdown-toggle" id="language-selector-toggle-btn">  <button tabindex="0" class="sui-button sm default outlined">   <svg style="min-width: 16px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" height="16" width="16" viewbox="0 0 24 24"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="m10.5 21l5.25-11.25L21 21m-9-3h7.5M3 5.621a49 49 0 0 1 6-.371m0 0q1.681 0 3.334.114M9 5.25V3m3.334 2.364C11.176 10.658 7.69 15.08 3 17.502m9.334-12.138q1.344.092 2.666.257m-4.589 8.495a18 18 0 0 1-3.827-5.802"/></svg>   </button>  </div> <ul class="sui-dropdown start" role="listbox" id="language-selector-dropdown" aria-labelledby="language-selector-toggle-btn" style="--offset: 0px;"> <li data-value="de" class="sui-dropdown-option has-icon" role="option" aria-selected="false"> <div class="sui-dropdown-line-container"> <svg style="min-width: 24px" xmlns="http://www.w3.org/2000/svg" prefix="lang-flags" height="24" width="24" viewbox="0 0 512 512"><mask id="IconifyId1999f87bd9397e8960"><circle cx="256" cy="256" r="256" fill="#fff"/></mask><g mask="url(#IconifyId1999f87bd9397e8960)"><path fill="#ffda44" d="m0 345l256.7-25.5L512 345v167H0z"/><path fill="#d80027" d="m0 167l255-23l257 23v178H0z"/><path fill="#333" d="M0 0h512v167H0z"/></g></svg> <span>Deutsch (de)</span> </div> </li><li data-value="en" class="sui-dropdown-option has-icon" role="option" aria-selected="false"> <div class="sui-dropdown-line-container"> <svg style="min-width: 24px" xmlns="http://www.w3.org/2000/svg" prefix="lang-flags" height="24" width="24" viewbox="0 0 512 512"><mask id="IconifyId1999f87bd9397e8961"><circle cx="256" cy="256" r="256" fill="#fff"/></mask><g mask="url(#IconifyId1999f87bd9397e8961)"><path fill="#eee" d="M256 0L0 256v64l32 32l-32 32v128l22-8l23 8h23l54-32l54 32h32l48-32l48 32h32l54-32l54 32h68l-8-22l8-23v-23l-32-54l32-54v-32l-32-48l32-48v-32l-32-54l32-54V0z"/><path fill="#d80027" d="M224 64v64h160l64-64zm0 128l32 64l-48 48v208h96V304h208v-96H304l16-16zM0 320v64h128l-64 64H0v64h45l131-131v-45l16-16zm336 16l176 176v-45L381 336Z"/><path fill="#0052b4" d="M0 0v256h256V0zm512 68L404 176h108zM404 336l108 108V336zm-228 68L68 512h108zm160 0v108h108z"/><path fill="#eee" d="m187 243l57-41h-70l57 41l-22-67zm-81 0l57-41H93l57 41l-22-67zm-81 0l57-41H12l57 41l-22-67zm162-81l57-41h-70l57 41l-22-67zm-81 0l57-41H93l57 41l-22-67zm-81 0l57-41H12l57 41l-22-67Zm162-82l57-41h-70l57 41l-22-67zm-81 0l57-41H93l57 41l-22-67Zm-81 0l57-41H12l57 41l-22-67Z"/></g></svg> <span>English (en)</span> </div> </li><li data-value="es" class="sui-dropdown-option has-icon" role="option" aria-selected="false"> <div class="sui-dropdown-line-container"> <svg style="min-width: 24px" xmlns="http://www.w3.org/2000/svg" prefix="lang-flags" height="24" width="24" viewbox="0 0 512 512"><mask id="IconifyId1999f87bd9397e8962"><circle cx="256" cy="256" r="256" fill="#fff"/></mask><g mask="url(#IconifyId1999f87bd9397e8962)"><path fill="#ffda44" d="m0 128l192-32l192 32l-256 256l-64 32l-64-32Z"/><rect width="16" height="24" x="128" y="192" fill="#d80027" rx="8"/><rect width="16" height="24" x="272" y="192" fill="#d80027" rx="8"/><rect width="32" height="16" x="120" y="208" fill="#ff9811" ry="8"/><rect width="32" height="16" x="120" y="304" fill="#ff9811" rx="8"/><path fill="#eee" d="M208 264a48 32 0 0 0-48 32a48 32 0 0 0 27.2 28.8L240 272a48 32 0 0 0-32-8m-80-40v80h16v-80zm144 0v16l16-16z"/><path fill="#d80027" d="M122 272a4 4 0 0 0-4 4a4 4 0 0 0 4 4h28a4 4 0 0 0 4-4a4 4 0 0 0-4-4z"/><path fill="#eee" d="M196 168c-7 0-13 5-15 11l-5-1c-9 0-16 7-16 16s7 16 16 16c7 0 13-4 15-11a16 16 0 0 0 17-4a16 16 0 0 0 17 4a16 16 0 1 0 10-20a16 16 0 0 0-27-5q-4.5-6-12-6m0 8c5 0 8 4 8 8c0 5-3 8-8 8c-4 0-8-3-8-8c0-4 4-8 8-8m24 0c5 0 8 4 8 8c0 5-3 8-8 8c-4 0-8-3-8-8c0-4 4-8 8-8m-44 10l4 1l4 8c0 4-4 7-8 7s-8-3-8-8c0-4 4-8 8-8m64 0c5 0 8 4 8 8c0 5-3 8-8 8c-4 0-8-3-8-7l4-8z"/><path fill="none" d="M220 284v12c0 7 5 12 12 12s12-5 12-12v-12z"/><path fill="#ff9811" d="M200 160h16v32h-16z"/><path fill="#d80027" d="m248 208l-8 8h-64l-8-8c0-13 18-24 40-24s40 11 40 24m-88 16h48v48h-48z"/><path fill="#d80027" d="M122 248a4 4 0 0 0-4 4a4 4 0 0 0 4 4h134l8-8z"/><path fill="#eee" d="M208 224v48h32l16-16v-32z"/><rect width="20" height="32" x="222" y="232" fill="#d80027" rx="10" ry="10"/><path fill="#ff9811" d="M168 232v8h8v16h-8v8h32v-8h-8v-16h8v-8zm8-16h64v8h-64z"/><g fill="#ffda44"><circle cx="186" cy="202" r="6"/><circle cx="208" cy="202" r="6"/><circle cx="230" cy="202" r="6"/></g><rect width="32" height="16" x="120" y="320" fill="#338af3" ry="8"/><path fill="#d80027" d="M0 384v128l128-128ZM0 0v128h384L512 0Z"/><path fill="#eee" d="m368 512l32-184l-32-184.5l-224 224l-32 72.5l32 72z"/><path fill="#d80027" d="M511.5 0L368 143.5V512h143.5z"/><path fill="#496e2d" d="M144 367.5L0 511.5v.5h144z"/><path fill="#ffda44" d="M256 277v10h12l10-22z"/><path fill="#338af3" d="M256 316c-14 0-28-5-40-13l6-9c20 13 48 13 68 0l7 9c-12 8-26 13-41 13"/><path fill="#751a46" d="m312.9 198.6l-30 30L301 239c4 11 12 21 23 26c3.5-32-2-52.8-11.1-66.4m-37 37l-24.8 24.8l9.9-2.4l18 18v-11l11 11l23-11l-35-21z"/><path fill="#496e2d" d="M341 242a85 85 0 0 1-137.8 66.3l-7.9 7.9A96 96 0 0 0 352 242Zm-28 17.5c-1.5 0-3 .5-4 1.5a78 78 0 0 1-53 21c-7.7 0-15.3-1.3-22.8-3.7l-8.6 8.6A96 96 0 0 0 250 293v17h12v-17c20-2 39-10 54-24c3-2 3-6 1-8c-1-1-2.5-1.5-4-1.5"/><circle cx="216" cy="323" r="8" fill="#ffda44"/><circle cx="297" cy="323" r="8" fill="#ffda44"/><circle cx="324" cy="302" r="8" fill="#ffda44"/><circle cx="341" cy="275" r="8" fill="#ffda44"/><path fill="#d80027" d="M208 272v24a24 24 0 0 0 1 7l31-31z"/><path fill="#ff9811" d="M272 208a8 8 0 1 0 0 16h16l8-8a8 8 0 0 0-8-8zm-112 64v24c0 8 4 14 9 19l5-6l5 10a21 21 0 0 0 10 0l5-10l4 5l8-8c1.3-3 2-6.3 2-10v-24h-9l-5 8l-5-8h-10l-5 8l-5-8z"/><path fill="#d80027" d="M169 272v43a24 24 0 0 0 10 4v-47zm20 0v47a24 24 0 0 0 5.6-1.6l4.4-4.4v-41z"/><circle cx="208" cy="272" r="16" fill="#338af3"/><rect width="34" height="22" x="239" y="299" fill="#ff9811" rx="11" ry="11"/></g></svg> <span>Español (es)</span> </div> </li><li data-value="fr" class="sui-dropdown-option has-icon" role="option" aria-selected="false"> <div class="sui-dropdown-line-container"> <svg style="min-width: 24px" xmlns="http://www.w3.org/2000/svg" prefix="lang-flags" height="24" width="24" viewbox="0 0 512 512"><mask id="IconifyId1999f87bd9397e8963"><circle cx="256" cy="256" r="256" fill="#fff"/></mask><g mask="url(#IconifyId1999f87bd9397e8963)"><path fill="#eee" d="M167 0h178l25.9 252.3L345 512H167l-29.8-253.4z"/><path fill="#0052b4" d="M0 0h167v512H0z"/><path fill="#d80027" d="M345 0h167v512H345z"/></g></svg> <span>Français (fr)</span> </div> </li><li data-value="zh" class="sui-dropdown-option has-icon" role="option" aria-selected="false"> <div class="sui-dropdown-line-container"> <svg style="min-width: 24px" xmlns="http://www.w3.org/2000/svg" prefix="lang-flags" height="24" width="24" viewbox="0 0 512 512"><mask id="IconifyId1999f87bd9397e8964"><circle cx="256" cy="256" r="256" fill="#fff"/></mask><g mask="url(#IconifyId1999f87bd9397e8964)"><path fill="#d80027" d="M0 0h512v512H0z"/><path fill="#ffda44" d="m140.1 155.8l22.1 68h71.5l-57.8 42.1l22.1 68l-57.9-42l-57.9 42l22.2-68l-57.9-42.1H118zm163.4 240.7l-16.9-20.8l-25 9.7l14.5-22.5l-16.9-20.9l25.9 6.9l14.6-22.5l1.4 26.8l26 6.9l-25.1 9.6zm33.6-61l8-25.6l-21.9-15.5l26.8-.4l7.9-25.6l8.7 25.4l26.8-.3l-21.5 16l8.6 25.4l-21.9-15.5zm45.3-147.6L370.6 212l19.2 18.7l-26.5-3.8l-11.8 24l-4.6-26.4l-26.6-3.8l23.8-12.5l-4.6-26.5l19.2 18.7zm-78.2-73l-2 26.7l24.9 10.1l-26.1 6.4l-1.9 26.8l-14.1-22.8l-26.1 6.4l17.3-20.5l-14.2-22.7l24.9 10.1z"/></g></svg> <span>简体中文（zh)</span> </div> </li> </ul> </div> <script type="module" src="/mock/path/LanguageSelector.astro?astro&type=script&index=0&lang.ts"></script> "`;
